torch/csrc/jit/tensorexpr/kernel.cpp

```diff
@@ -1591,6 +1602,107 @@ void TensorExprKernel::bindInput(const torch::jit::Value* input) {
   }
 }
 
+namespace {
+
+// Remove all indices from axes positions.
+std::vector<VarHandle> squeezeIndices(
+    const ParameterList& indices,
+    const std::vector<size_t>& axes) {
+  std::vector<VarHandle> indices_squeezed;
+  for (size_t dim = 0; dim < indices.size(); ++dim) {
+    if (!std::count(axes.begin(), axes.end(), dim)) {
+      indices_squeezed.push_back(indices[dim]);
+    }
+  }
+  return indices_squeezed;
+}
+
+} // namespace
+
+Tensor* TensorExprKernel::computeSum(const torch::jit::Value* v) {
+  auto reduction_info = getReductionInfo(v->node());
+  return Reduce(
+      "sum",
+      reduction_info.outputDims,
+      Sum(),
+      [&](ParameterList& indices) {
+        const auto& axes = reduction_info.axes;
+        // "Squeeze" out indices inserted when keepdim is set.
+        auto indices_squeezed =
+            reduction_info.keepdim ? squeezeIndices(indices, axes) : indices;
+        TORCH_INTERNAL_ASSERT(axes.size() <= indices_squeezed.size());
+        // Move innermost indices into axes positions:
+        //   1. Fill the outermost indices first.
+        //   2. Insert the innermost indices into the correct axis position,
+        //   displacing the outermost indices as needed.
+        std::vector<ExprHandle> indices_exprs;
+        size_t i = 0;
+        for (; i < indices_squeezed.size() - axes.size(); ++i) {
+          indices_exprs.push_back(indices_squeezed[i]);
+        }
+        for (auto axis : axes) {
+          indices_exprs.insert(
+              indices_exprs.begin() + axis, indices_squeezed[i]);
+          ++i;
+        }
+        auto indexed = tensorOrConstant(v->node()->input(0), indices_exprs);
+        if (reduction_info.dtype) {
+          return Cast::make(*reduction_info.dtype, indexed);
+        } else {
+          return indexed;
+        }
+      },
+      reduction_info.reductionDims);
+}
+
+TensorExprKernel::ReductionInfo TensorExprKernel::getReductionInfo(
+    const torch::jit::Node* node) {
+  std::vector<size_t> axes;
+  bool keepdim = false;
+  auto sizes = sizesForValue(node->namedInput(attr::self));
```

Comment:
Hm, that's a little surprising, what IR do you expect to see here? When I'm inspecting IR for `x.sum()` it doesn't have any attributes:
```
import torch

def foo(x):
    return x.sum()

f = torch.jit.script(foo)
print(f.graph)
```
The corresponding IR:
```
graph(%x.1 : Tensor):
  %2 : None = prim::Constant()
  %3 : Tensor = aten::sum(%x.1, %2) # u.py:4:11
  return (%3)
```

What node are we calling `namedInput` on?

In either case, I think it could be useful to show the expected IR in a comment.

Reply:
Actually, nvm, I didn't know we had such a fancy API for accessing inputs by their name in schema :) Showing the schema here could still be helpful.

Reply:
I've added a comment, not sure what's the confusion here or whether the expected IR would improve matters. The `aten::sum`  operator has a named input `self`, per schema: https://github.com/pytorch/pytorch/blob/0521c712410ae014844d680cfc40b59c6f456bcf/aten/src/ATen/native/native_functions.yaml#L2939. Is there a different / better way to do this? The `attr::` namespace is a bit confusing, since it's not an actual attribute.

Reply:
I usually just referenced inputs by their index (i.e. `n->input(2)`) - but I think addressing them by name makes it a bit easier to read despite they are not really attributes.